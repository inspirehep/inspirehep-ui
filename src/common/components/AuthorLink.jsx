import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import { Tooltip } from 'antd';

import ExternalLink from './ExternalLink';
import AuthorAffiliationList from './AuthorAffiliationList';

class AuthorLink extends Component {
  getAuthorHref() {
    const { author, recordId } = this.props;
    let href = `//inspirehep.net/author/profile/${author.get('full_name')}`;
    if (recordId != null) {
      href = `${href}?recid=${recordId}`;
    }
    return href;
  }

  getFullName() {
    const { author } = this.props;
    if (author.has('first_name')) {
      const firstName = author.get('first_name');
      const lastName = author.get('last_name', '');
      return `${firstName} ${lastName}`;
    }
    return author.get('full_name');
  }

  renderEditorSuffix() {
    const { author } = this.props;
    const roles = author.get('inspire_roles', []);

    if (roles.includes('editor')) {
      return <Tooltip title="editor">(ed.)</Tooltip>;
    }
    return null;
  }

  renderAffiliationsList() {
    const { author } = this.props;
    const affiliations = author.get('affiliations');
    return (
      affiliations && (
        <span className="pl1">
          (
          <AuthorAffiliationList
            affiliations={affiliations.map(affiliation =>
              affiliation.get('value')
            )}
          />
          )
        </span>
      )
    );
  }

  render() {
    const authorHref = this.getAuthorHref();
    return (
      <div className="di">
        <ExternalLink href={authorHref}>{this.getFullName()}</ExternalLink>
        {this.renderAffiliationsList()}
        {this.renderEditorSuffix()}
      </div>
    );
  }
}

AuthorLink.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
  recordId: PropTypes.number,
};

AuthorLink.defaultProps = {
  recordId: undefined,
};

export default AuthorLink;
