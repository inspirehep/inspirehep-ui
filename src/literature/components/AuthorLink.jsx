import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';

class AuthorLink extends Component {
  getAuthorHref() {
    const { author, recordId } = this.props;
    let href = `//inspirehep.net/author/profile/${author.get('full_name')}`;
    if (recordId != null) {
      href = `${href}?recid=${recordId}`;
    }
    return href;
  }

  getAffiliationHref() {
    const { author } = this.props;
    const affiliation = author.getIn(['affiliations', 0, 'value']);
    return `//inspirehep.net/search?cc=Institutions&p=institution:"${affiliation}"`;
  }

  getFullName() {
    const { author } = this.props;
    if (author.has('first_name')) {
      const firstName = author.get('first_name');
      const lastName = author.get('last_name', '');
      return `${firstName} ${lastName}`;
    }
    return author.get('full_name');
  }

  renderAffiliationLink() {
    const { author } = this.props;
    const affiliation = author.getIn(['affiliations', 0, 'value']);
    if (affiliation) {
      const affiliationHref = this.getAffiliationHref();
      return (
        <span className="pl1 grey">
          (
          <a target="_blank" href={affiliationHref}>
            {affiliation}
          </a>
          )
        </span>
      );
    }
    return null;
  }

  render() {
    const authorHref = this.getAuthorHref();

    return (
      <div className="di">
        <a target="_blank" href={authorHref}>
          {this.getFullName()}
        </a>
        {this.renderAffiliationLink()}
      </div>
    );
  }
}

AuthorLink.propTypes = {
  author: PropTypes.instanceOf(Map).isRequired,
  recordId: PropTypes.number,
};

AuthorLink.defaultProps = {
  recordId: undefined,
};

export default AuthorLink;
