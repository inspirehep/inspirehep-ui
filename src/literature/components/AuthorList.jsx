import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import { Modal } from 'antd';

import InlineList from '../../common/components/InlineList';
import AuthorLink from './AuthorLink';

class AuthorList extends Component {
  constructor(props) {
    super(props);
    this.onModalCancel = this.onModalCancel.bind(this);
    this.onModalOpen = this.onModalOpen.bind(this);

    this.state = {
      modalVisible: false,
    };
  }

  onModalOpen() {
    this.setState({ modalVisible: true });
  }

  onModalCancel() {
    this.setState({ modalVisible: false });
  }

  renderSuffix() {
    const { showAll } = this.props;
    if (showAll) {
      return (
        <div className="di">
          {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions, jsx-a11y/anchor-is-valid */}
          <a onClick={this.onModalOpen}> Show all</a>
        </div>
      );
    }
    return <span> et al.</span>;
  }

  renderInlineList(items, suffix = true) {
    const { authors, limit, recordId, wrapperClassName } = this.props;
    return (
      <InlineList
        wrapperClassName={wrapperClassName}
        items={items}
        suffix={authors.size > limit && suffix ? this.renderSuffix() : null}
        extractKey={author => author.get('full_name')}
        renderItem={author => (
          <AuthorLink author={author} recordId={recordId} />
        )}
      />
    );
  }

  render() {
    const { modalVisible } = this.state;
    const { authors, limit, total } = this.props;
    const showTotal = total === -1 ? authors.size : total;
    return (
      <Fragment>
        {this.renderInlineList(authors.take(limit))}
        <Modal
          title={`${showTotal} authors`}
          width="50%"
          visible={modalVisible}
          footer={null}
          onCancel={this.onModalCancel}
        >
          {this.renderInlineList(authors, false)}
        </Modal>
      </Fragment>
    );
  }
}

AuthorList.propTypes = {
  authors: PropTypes.instanceOf(List),
  limit: PropTypes.number,
  recordId: PropTypes.number,
  showAll: PropTypes.bool,
  total: PropTypes.number,
  wrapperClassName: PropTypes.string,
};

AuthorList.defaultProps = {
  authors: List(),
  limit: 5,
  recordId: undefined,
  showAll: false,
  total: -1,
  wrapperClassName: null,
};

export default AuthorList;
