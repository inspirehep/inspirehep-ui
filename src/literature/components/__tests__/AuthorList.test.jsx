import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import AuthorList from '../AuthorList';
import AuthorLink from '../AuthorLink';

describe('AuthorList', () => {
  it('renders only 5 authors and suffixes "show all" if passed more', () => {
    const authors = fromJS([
      {
        full_name: 'Test, Guy 1',
      },
      {
        full_name: 'Test, Guy 2',
      },
      {
        full_name: 'Test, Guy 3',
      },
      {
        full_name: 'Test, Guy 4',
      },
      {
        full_name: 'Test, Guy 5',
      },
      {
        full_name: 'Test, Guy 6',
      },
    ]);
    const wrapper = shallow(<AuthorList recordId={12345} authors={authors} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders only limited (prop) authors and suffixes "show all" if passed more', () => {
    const authors = fromJS([
      {
        full_name: 'Test, Guy 1',
      },
      {
        full_name: 'Test, Guy 2',
      },
      {
        full_name: 'Test, Guy 3',
      },
    ]);
    const wrapper = shallow(
      <AuthorList limit={2} recordId={12345} authors={authors} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders all authors if they are less than the limit wihtout "show all."', () => {
    const authors = fromJS([
      {
        full_name: 'Test, Guy 1',
      },
      {
        full_name: 'Test, Guy 2',
      },
    ]);
    const wrapper = shallow(
      <AuthorList limit={4} recordId={12345} authors={authors} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders authors by using AuthorLink', () => {
    const authors = fromJS([
      {
        full_name: 'Test, Guy 1',
      },
    ]);
    const wrapper = shallow(
      <AuthorList limit={4} recordId={12345} authors={authors} />
    );

    // Fragment make the ``dive`` a bit more difficult
    expect(
      wrapper
        .find('InlineList')
        .first()
        .dive()
        .find(AuthorLink).length
    ).toBe(1);
  });
});
