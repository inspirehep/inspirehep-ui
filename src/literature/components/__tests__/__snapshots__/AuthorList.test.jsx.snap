// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorList renders all authors if they are less than the limit wihtout "et al." 1`] = `
<div
  className="__InlineList__"
>
  <ul
    className="separate-items"
  >
    <li
      key="Test, Guy 1"
    >
      <AuthorLink
        author={
          Immutable.Map {
            full_name: "Test, Guy 1",
          }
        }
        recordId={12345}
      />
    </li>
    <li
      key="Test, Guy 2"
    >
      <AuthorLink
        author={
          Immutable.Map {
            full_name: "Test, Guy 2",
          }
        }
        recordId={12345}
      />
    </li>
  </ul>
</div>
`;

exports[`AuthorList renders only 5 authors and suffixes "et al." if passed more 1`] = `
<InlineList
  extractKey={[Function]}
  items={
    Immutable.List [
      Immutable.Map {
        full_name: "Test, Guy 1",
      },
      Immutable.Map {
        full_name: "Test, Guy 2",
      },
      Immutable.Map {
        full_name: "Test, Guy 3",
      },
      Immutable.Map {
        full_name: "Test, Guy 4",
      },
      Immutable.Map {
        full_name: "Test, Guy 5",
      },
    ]
  }
  label={null}
  renderItem={[Function]}
  separateItems={true}
  suffix={
    <span>
       et al.
    </span>
  }
  wrapperClassName={null}
/>
`;

exports[`AuthorList renders only limited (prop) authors and suffixes "et al." if passed more 1`] = `
<div
  className="__InlineList__"
>
  <ul
    className="separate-items"
  >
    <li
      key="Test, Guy 1"
    >
      <AuthorLink
        author={
          Immutable.Map {
            full_name: "Test, Guy 1",
          }
        }
        recordId={12345}
      />
    </li>
    <li
      key="Test, Guy 2"
    >
      <AuthorLink
        author={
          Immutable.Map {
            full_name: "Test, Guy 2",
          }
        }
        recordId={12345}
      />
    </li>
  </ul>
  <span>
     et al.
  </span>
</div>
`;
